### PR TITLE
CI: Migrate from Travis to GitHub Actions

### PR DESCRIPTION
**Pull request by HarshCasper**
## Description

This PR:

- Deletes the existing `.travis.yml`
- Adds a GHA job to build and test
- Adds a GHA step for uploading coverage to Codecov
- Runs CI on Python `3.7`, `3.8`, `3.9`, `3.10`
- Builds and tests documentation on Sphinx

### CODE REVIEW

**Comment by czgdp1807** [comment]
The codecov bot hasn't made a comment.

File: .github/workflows/ci.yml [hunk 1]
```diff
@@ -0,0 +1,62 @@
+name: Python CI Workflow
```
Comment (czgdp1807) [inline .github/workflows/ci.yml:1]:
```suggestion
name: PyDataStructs Continuous Integration Tests.
```

File: .github/workflows/ci.yml [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+
+on:
+  push:
+    branches: [master]
+  pull_request:
+    branches: [master]
+
+jobs:
+  test:
+    runs-on: ubuntu-latest
```
Comment (czgdp1807) [inline .github/workflows/ci.yml:11]: Is windows available as an option?

Comment (HarshCasper) [reply .github/workflows/ci.yml:11]: Yes, it is. Do you want to run a matrix configuration for all OSes?

Comment (czgdp1807) [reply .github/workflows/ci.yml:11]: If we can run for free then why not?😉

File: .github/workflows/ci.yml [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+          python-version: ${{ matrix.python-version }}
+
+      - name: Upgrade pip version
+        run: |
+          python3 -m pip install --upgrade pip
+
+      - name: Install requirements
+        run: |
+          pip3 install -r requirements.txt
+          pip3 install -r docs/requirements.txt
```
Comment (czgdp1807) [inline .github/workflows/ci.yml:36]: Does the following work?
```suggestion
          pip install -r requirements.txt
          pip install -r docs/requirements.txt
```

File: .github/workflows/ci.yml [hunk 4]
```diff
@@ -0,0 +1,62 @@
...
+      - uses: actions/checkout@v2
+
+      - name: Set up Python ${{ matrix.python-version }}
+        uses: actions/setup-python@v2
+        with:
+          python-version: ${{ matrix.python-version }}
+
+      - name: Upgrade pip version
+        run: |
+          python3 -m pip install --upgrade pip
```
Comment (czgdp1807) [inline .github/workflows/ci.yml:31]: Similarly, the following,
```suggestion
          python -m pip install --upgrade pip
```

**Comment by HarshCasper** [comment]
@czgdp1807 The Windows and macOS tests were failing because of an action that I was using to build docs that was only compatible with Linux. The CI should work fine now.

**Comment by HarshCasper** [comment]
@czgdp1807 It works :smile:

**Comment by czgdp1807** [comment]
Yeah. Great work. Is there anything we can do to add a timeout using secrets? Let me know if I need to set an environment variable in my Github account.

**Comment by HarshCasper** [comment]
Hi @czgdp1807 

I have added a timeout secret. To set up a secret, go to your Repository Settings page, then select Secrets. You can add a secret named `TIMEOUT_MINUTES` and provide a strict numerical value in minutes denoting the maximum time for which the workflow should run.

You can kick-start the CI once again and we will see if its working as per our expectations :)

**Comment by czgdp1807** [comment]
Tests don't seem to start. I have added the secret though.

File: .github/workflows/ci.yml [hunk 5]
```diff
@@ -0,0 +1,63 @@
...
+on:
+  push:
+    branches: [master]
+  pull_request:
+    branches: [master]
+
+jobs:
+  test:
+    runs-on: ${{matrix.os}}
+    timeout-minutes: ${{secret.TIMEOUT_MINUTES}}
```
Comment (czgdp1807) [inline .github/workflows/ci.yml:12]:
```suggestion
    timeout-minutes: ${{secrets.TIMEOUT_MINUTES}}
```

File: .github/workflows/ci.yml [hunk 6]
```diff
@@ -0,0 +1,63 @@
...
+on:
+  push:
+    branches: [master]
+  pull_request:
+    branches: [master]
+
+jobs:
+  test:
+    runs-on: ${{matrix.os}}
+    timeout-minutes: ${{ secrets.TIMEOUT_MINUTES }}
```
Comment (czgdp1807) [inline .github/workflows/ci.yml:12]:
```suggestion
    timeout-minutes: ${{ TIMEOUT_MINUTES }}
```

**Comment by czgdp1807** [comment]
The error,
```
Invalid workflow file : .github/workflows/ci.yml#L12
The workflow is not valid. .github/workflows/ci.yml (Line: 12, Col: 22): Unrecognized named-value: 'TIMEOUT_MINUTES'. Located at position 1 within expression: TIMEOUT_MINUTES .github/workflows/ci.yml (Line: 12, Col: 22): Unexpected value '${{ TIMEOUT_MINUTES }}'
```

**Comment by HarshCasper** [comment]
@czgdp1807 Seems like we cannot add a secret as a job value. We have to define it manually here.

**Comment by czgdp1807** [comment]
Okay. Lets do that for now.

**Comment by HarshCasper** [comment]
Hi @czgdp1807 

Apologies for the delay. I have set the timeout minutes. Let me know if you wish to have further changes or need the commits to be squashed before merging. Thanks!

File: .github/workflows/ci.yml [hunk 7]
```diff
@@ -0,0 +1,63 @@
...
+on:
+  push:
+    branches: [master]
+  pull_request:
+    branches: [master]
+
+jobs:
+  test:
+    runs-on: ${{matrix.os}}
+    timeout-minutes: 2
```
Comment (czgdp1807) [inline .github/workflows/ci.yml:12]:
```suggestion
    timeout-minutes: 5
```

**Comment by czgdp1807** [comment]
Thanks @HarshCasper for this.

**Comment by czgdp1807** [comment]
The badge in the README needs to be updated though. As of now it uses to Travis CI for showing, `build:passing`. It should use Github Actions now.

**Comment by HarshCasper** [comment]
I will make a follow-up PR for that. Should I also add Flake8 and format code using Black? It would be a comparatively larger PR and I would like to know if you would be okay with that?

**Comment by czgdp1807** [comment]
> Should I also add Flake8 and format code using Black?

N.B. - https://github.com/codezonediitj/pydatastructs/pull/324#issuecomment-948802930

**Pull request merged**